### DISCUSSION TITLE
IPv8 Stream Transport fun

### DISCUSSION THREAD

**Discussion by hbiyik** [Tribler/tribler#5656] [Q&A]
I am having some fun with IPv8 library. My personal project for fun aims to realtime stream transport P2P, no multicast no anycast.
I think i got the hang of it IPv8 i was somehow familiar with dispersy, but for me the question is there any available communities, that has stream transport support. Like sequencecontrol, and may be window size like bandwidth control builtins? I could not see such stuff in tunnel community

For future may be i can try to some fun with TLS like end 2 end encryption but yeah i think fist i need a transport.

**Comment by egbertbouman** [comment]
There is no streaming support in IPv8. Mostly because we currently don't need it. The `TunnelCommunity` also doesn't have it, because it relies on the BitTorrent protocol to deal with things like bandwidth control.

In the past, I've played around with adding `libutp` to IPv8 in order to do bulk data transfer. Unfortunately, I can't seem to find the implementation. Anyway, that could be an option.

**Comment by hbiyik** [comment]
ok got it, thanks. let me fiddle around may be i can make sth interesting out of this experiment this is also good practive to get cosy with ipv8.

**Comment by qstokkink** [comment]
I still have my uTP over Dispersy community lying around if you need inspiration: https://github.com/qstokkink/gumby/blob/utp_community/experiments/utp/utp_community.py That would have to be ported to IPv8 though (which would shed quite a bit of code).

- I dropped this idea as I was remaking a large chunk of libtorrent instead of.. just using libtorrent. Less code, less maintenance.

I also have some code for encrypting all communication by default: https://github.com/qstokkink/py-ipv8/tree/default_e2e_enc

- I dropped that idea in favor of routing over the TunnelCommunity (all overlays now take an `anonymize=True` argument to do so easily).

Beyond what @egbertbouman already mentioned: these features are not (and probably should not get) in the core IPv8 protocol as they are not part of the core mission of IPv8. IPv8 is for peer-to-peer overlay messaging instead of peer-to-peer file sharing.

That said, it is a lot of fun to make these things.

**Comment by synctext** [comment]
We worked with several people on live streaming and got it working together with BBC and Pioneer hardware team from UK:
https://github.com/libswift/libswift
We even managed to run our streaming engine ***inside*** a Samsung television with rooted firmware.
Lesson: it needs a team of 100+ engineers to move from 99% works to 99.9% works.
![RebellionTV_screenshot](https://user-images.githubusercontent.com/325224/96442245-16b78c00-120b-11eb-924b-4a90f973e7bc.jpg)

**Comment by hbiyik** [comment]
@qstokkink those stuff are golden to me but i will try to reinvent the wheel, and probably copy some of stuff from your codes as well specially what you did in e2e i find it very interesting.
@synctext i remember that, i think there was also a plugin called P2Pnextplayer or swarmplayer it was..

**Reply by hbiyik** [reply]
I also had a feeling that later aceplayer was a ripoff from libswift but never had any proof or any serious investigation i had done on it, i think it will always stay as my personal conspiracy theory :)
